add technote about combining data from various sources into a single field (#267)
apollographql/docs: src/content/technotes/TN0018-aggregating-across-subgraphs.mdx

id: TN0018
tags: [federation, best practices]
---

Apollo Federation is great for orchestrating data fetches across multiple domains, but it does not help with several common problems in distributed architectures:

- Searching across a variety of types stored in multiple data sources
- Combining list of multiple types from multiple sources into a single list, especially when paged
- Filtering a list based on attributes that exist in various data sources
- Deriving data from an aggregate of multiple data sources

While it is sometimes possible to bend the federation query planner to support these use cases using the `@requires` directive, it's almost always better to provide this functionality in a new system such as a search index.

## An example of the problem: search

Given an operation to search across both books and movies, we want to return a polymorphic list of books and movies that match the search term.

```graphql
query SearchEverything($query: String!) {
  search(query: $query) {
    ... on Book {
      title
      authors {
        name
      }
    }
    ... on Movie {
      title
      directors {
        name
      }
    }
  }
}
```

> If the `Book` type and `Movie` type are defined in and provided by different subgraphs, the question is: who provides the `Query.search` root field?

Author: hmm you prefer the passive voice here?

> If the `Book` type and `Movie` type are defined in and provided by different subgraphs, the question is: who provides the `Query.search` root field?

Author: whoops i flipped it what i was reading! *i'm* the one with passive voice. yikes

      {
        search(query: $query) {
          __typename
          ... on Book {
            __typename
            id
            title
          }
          ... on Movie {
            __typename
            id
            title
          }
        }
      }
    },
    Parallel {
      Flatten(path: "search.@") {
        Fetch(service: "one") {
          {
            ... on Book {
              __typename
              id
            }
          } =>
          {
            ... on Book {
              authors {
                name
              }
            }
          }
        },
      },
      Flatten(path: "search.@") {
        Fetch(service: "two") {
          {
            ... on Movie {
              __typename
              id
            }
          } =>
          {
            ... on Movie {
              directors {
                name
              }
            }
          }
        },
      },
    },
  },
}
```

</ExpansionPanel>

Note that the **Search** subgraph provides a minimal set of fields. We don't need to duplicate the entire `Book` and `Movie` types in our **Search** domain, just the fields we want to search on.

### Tradeoffs

> Adding a whole new domain to your supergraph is not without downsides, of course.

Reviewer: ```suggestion
As you'd expect, adding an entirely new domain to your supergraph has its tradeoffs:
```

#### The question of ownership

> In the **Search** example, we've introduced multiple new services that require ongoing development, maintenance, and support. It doesn't make sense for our existing Books and Movies teams to take on this extra burden. Usually we want to spin up a whole new team to hold the pager and deployment keys for our new subgraphs and data stores.

Reviewer: ```suggestion
In the Search subgraph example, we've introduced multiple new services that require ongoing development, maintenance, and support. It doesn't make sense for our existing Books and Movies teams to take on this extra burden. Usually, we want to spin up an entirely new team to hold the pager and deployment keys for our new subgraphs and data stores.
```

#### Eventual consistency

> Replicating data from the canonical Books and Movies databases into the Search Index inevitably involves replication lag, leading to eventually consistent results between our subgraphs.

Reviewer: ```suggestion
Replicating data from the canonical Books and Movies databases into the Search index inevitably involves replication lag, leading to eventually consistent results between our subgraphs.
```

> How we handle that lag depends on our business requirements. If the results from `Query.search` must be "internally" consistent, we can denormalize data using `@shareable` and `@provides`. We demonstrated this by providing the `title` fields from the **Search** subgraph and its backing index.

Reviewer: ```suggestion
How we handle that lag depends on our business requirements. If the results from `Query.search` must be "internally" consistent, we can denormalize data using `@shareable` and `@provides`. We demonstrated this by providing the `title` fields from the Search subgraph and its backing index.
```

> If the results must be accurate with our canonical data source, we can have the query planner fetch those fields from their respective subgraphs. The `Book.authors` and `Movie.directors` fields exemplify that pattern.

Reviewer: ```suggestion
If the results must be accurate with our canonical data source, we can make sure the query planner fetches those fields from their respective subgraphs. The `Book.authors` and `Movie.directors` fields exemplify that pattern.
```
